Set appArmor profile for istio-cni to unconfined

jaellio opened the pull request:
**Please provide a description of this PR:**

- Conditionally set CAP_SYS_PTRACE when repairPods is enabled (REMOVED)
- Set appArmor profile for istio-cni to unconfined

jaellio committed:
Set appArmor profile for istio-cni to unconfined

Conditionally set CAP_SYS_PTRACE when repairPods is enabled

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

istio-testing commented:
Skipping CI for Draft Pull Request.
If you want CI signal for your change, please convert it to an actual PR.
You can still manually trigger a test run with `/test all`

jaellio commented:
/test all

istio-testing commented:
@jaellio: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
release-notes_istio | 2d8275c5dcc8e1681adc6317281cc4572eabc8cb | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/53888/release-notes_istio/1856888712911130624) | true | `/test release-notes`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

jaellio committed:
Add release note

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

jaellio commented:
Still looking into whether or not PTRACE is needed for ambient or just repair mode. Want to confirm the ambient and repair mode disabled scenario is covered in testing.

zirain commented on manifests/charts/istio-cni/cni-1.0.0.tgz:
what is this?

zirain commented on manifests/charts/istio-cni/templates/daemonset.yaml:
```diff
@@ -92,6 +92,8 @@ spec:
               path: /readyz
               port: 8000
           securityContext:
+            appArmorProfile:
```
is this the default value? if not, what about make it optional by `with values.appArmorProfile` or something else?

bleggett replied:
1. It is not the default value.
2. Because `istio-cni` is privileged, it is the only value that `istio-cni` can work with on systems with appArmor.
3. It is a no-op on any platform without appArmor.
4. So this value will work everywhere.
5. We have no reason given the above to expose and document it as a toggle to users, but if someone later comes up with a real need to do that, it would be easy enough to do at that time.

zirain replied:
what I'm thinking is that helm template will generated same manifests output with same `values.yaml` across different versions.

jaellio committed:
Remove unintentionally added file

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

bleggett commented on cni/pkg/nodeagent/ztunnelserver.go:
```diff
@@ -327,6 +327,7 @@ func (z *ztunnelServer) PodAdded(ctx context.Context, pod *v1.Pod, netns Netns)
 	if err != nil {
 		return err
 	}
+	log.Debug("sent pod add to ztunnel")
```
nit: if we keep this, make it a `trace`, we already have quite a bit of debug logging in the node agent.

howardjohn replied:
unfortunately there is no trace in our go logging

jaellio replied:
I mostly added this since it wasn't immediately clear in the logs that the request successfully sent. I'm fine with removing this.

zirain replied:
should we add pod name here to make it more meaningful?

jaellio replied:
The pod name is already included in the log here https://github.com/istio/istio/pull/53888/files#diff-9873062ab6cdec1c9e1829c5cad625c4ef9d878a05b8b96ee3449623ea50f442R312

bleggett commented on manifests/charts/istio-cni/templates/daemonset.yaml:
```diff
@@ -108,8 +110,10 @@ spec:
               # CAP_SYS_PTRACE is required for repair mode to describe the pod's network namespace
-              # in ambient and repair mode.
+              # in repair mode.
+{{- if .Values.repair.repairPods }}
```
What exact thing does `repair` do that ambient does not with network namespaces that requires this?

They are separate features but they do nearly identical things with pod netnses, so I'm slightly surprised this is only required for one of them.

jaellio replied:
When I was testing this change https://github.com/istio/istio/pull/53478 I got the following error when testing the repair controller: `failed to Statfs "/host/proc/1/ns/net": permission denied`. If the repair controller is not invoked or enabled I wasn't seeing this failure. I am digging into this more now.

bleggett replied:
Ok - we [definitely `stat` procfs netns FDs in ambient too](https://github.com/istio/istio/blob/50dc03e621d9a457df479f8fe5f7b8241878d45d/cni/pkg/nodeagent/podcgroupns.go#L117), so I'd expect it to fail in both spots.

jaellio replied:
Agree, working on writing tests now.

jaellio replied:
I modified the repair controller values set [here](https://github.com/istio/istio/blob/master/tests/integration/ambient/main_test.go#L112) for the ambient integration tests to include `repairPods: false`. Based on the above changes this would remove CAP_SYS_PTRACE from the istio-cni-node's capabilities. With repair disabled, repair pods explicitly disabled, and the exclusion on CAP_SYS_PTRACE the ambient traffic tests still pass. I didn't expect this... looking into it

jaellio replied:
Never mind, looks like there are some test failures. Will take a look.

jaellio replied:
I am not able to reproduce the test failures locally. Adding ptrace cap back for ambient for now

howardjohn reviewed (commented):
lgtm sans the ptracr question

jaellio committed:
Explicitly set repairPods to false when repair controller is
disabled for testing. Istio-cni-node daemonset configuration
is dependent on the value of repairPods. If repair is false,
repairPods should also be false.

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

istio-testing commented:
@jaellio: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
integ-ambient_istio | b172b26ea223cf0fdc083c6b6a6f1a2af07afa80 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/53888/integ-ambient_istio/1857202567537233920) | true | `/test integ-ambient`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

jaellio committed:
Add back ptrace cap for ambient

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

istio-testing commented:
@jaellio: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
integ-ambient_istio | b172b26ea223cf0fdc083c6b6a6f1a2af07afa80 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/53888/integ-ambient_istio/1857496198651318272) | true | `/test integ-ambient`
unit-tests-arm64_istio | f40f472818adc9fcd9b550bfdc3e901dd38ff025 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/53888/unit-tests-arm64_istio/1858643472542601216) | true | `/test unit-tests-arm64`
gencheck_istio | f40f472818adc9fcd9b550bfdc3e901dd38ff025 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/53888/gencheck_istio/1858643460513337344) | true | `/test gencheck`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@jaellio: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
unit-tests-arm64_istio | f40f472818adc9fcd9b550bfdc3e901dd38ff025 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/53888/unit-tests-arm64_istio/1858643472542601216) | true | `/test unit-tests-arm64`
gencheck_istio | f40f472818adc9fcd9b550bfdc3e901dd38ff025 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/53888/gencheck_istio/1858643460513337344) | true | `/test gencheck`
integ-ambient_istio | f40f472818adc9fcd9b550bfdc3e901dd38ff025 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/53888/integ-ambient_istio/1858643460551086080) | true | `/test integ-ambient`
integ-cni_istio | f40f472818adc9fcd9b550bfdc3e901dd38ff025 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/53888/integ-cni_istio/1858643460672720896) | true | `/test integ-cni`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@jaellio: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
unit-tests-arm64_istio | f40f472818adc9fcd9b550bfdc3e901dd38ff025 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/53888/unit-tests-arm64_istio/1858643472542601216) | true | `/test unit-tests-arm64`
gencheck_istio | f40f472818adc9fcd9b550bfdc3e901dd38ff025 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/53888/gencheck_istio/1858643460513337344) | true | `/test gencheck`
integ-ambient_istio | f40f472818adc9fcd9b550bfdc3e901dd38ff025 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/53888/integ-ambient_istio/1858643460551086080) | true | `/test integ-ambient`
integ-cni_istio | f40f472818adc9fcd9b550bfdc3e901dd38ff025 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/53888/integ-cni_istio/1858643460672720896) | true | `/test integ-cni`
lint_istio | f40f472818adc9fcd9b550bfdc3e901dd38ff025 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/53888/lint_istio/1858643468889362432) | true | `/test lint`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

howardjohn commented on manifests/charts/istio-cni/templates/daemonset.yaml:
```diff
@@ -107,8 +109,8 @@ spec:
-              # CAP_SYS_PTRACE is required for repair mode to describe the pod's network namespace
-              # in ambient and repair mode.
+              # CAP_SYS_PTRACE is required for repair and ambient mode to describe
+              $ the pod's network namespace.
```
```suggestion
              # the pod's network namespace.
```

howardjohn commented on tests/integration/ambient/main_test.go:
```diff
@@ -110,6 +110,7 @@ values:
     # Refer to issue #49207 for more context.
     repair:
       enabled: false
+      repairPods: false
```
nit: we can probably drop this?

howardjohn commented on releasenotes/notes/53829.yaml:
```diff
@@ -0,0 +1,11 @@
+  AppArmor profiles which block certain privileged pod capabilities. Previously, AppArmor
+  (when enabled) was bypassed for the istio-cni-node DaemonSet since privileged was set to true
+  in the SecurityContext. This change ensures that the AppArmor profile is set to unconfined
+  for the istio-cni-node DaemonSet. The CAP_SYS_PTRACE capability is now conditionally added
+  based on the value of `.Values.repair.repairPods`.
```
Outdated now

istio-testing commented:
@jaellio: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
unit-tests-arm64_istio | f40f472818adc9fcd9b550bfdc3e901dd38ff025 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/53888/unit-tests-arm64_istio/1858643472542601216) | true | `/test unit-tests-arm64`
gencheck_istio | f40f472818adc9fcd9b550bfdc3e901dd38ff025 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/53888/gencheck_istio/1858643460513337344) | true | `/test gencheck`
integ-ambient_istio | f40f472818adc9fcd9b550bfdc3e901dd38ff025 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/53888/integ-ambient_istio/1858643460551086080) | true | `/test integ-ambient`
integ-cni_istio | f40f472818adc9fcd9b550bfdc3e901dd38ff025 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/53888/integ-cni_istio/1858643460672720896) | true | `/test integ-cni`
lint_istio | f40f472818adc9fcd9b550bfdc3e901dd38ff025 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/53888/lint_istio/1858643468889362432) | true | `/test lint`
integ-helm_istio | f40f472818adc9fcd9b550bfdc3e901dd38ff025 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/53888/integ-helm_istio/1858643460836298752) | true | `/test integ-helm`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

jaellio committed:
Update release note, fix typo, nit

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

howardjohn commented:
Approved but it has a do-not-merge label

bleggett reviewed (approved):
SGTM, ty @jaellio!

istio-testing commented:
In response to a cherrypick label: new pull request created: #53947
